weld/jpa/src/main/java/org/jboss/as/weld/services/bootstrap/WeldJpaInjectionServices.java

```diff
@@ -128,4 +144,143 @@ public EntityManagerResourceReferenceFactory(String scopedPuName, EntityManagerF
             return new SimpleResourceReference<EntityManager>(result);
         }
     }
+
+    private static class LazyEntityManagerFactory<T> implements ResourceReferenceFactory<EntityManagerFactory> {
+        private final PersistenceUnitServiceImpl persistenceUnitService;
+        private final ServiceController<?> serviceController;
+        public LazyEntityManagerFactory(PersistenceUnitServiceImpl persistenceUnitService, ServiceController<?> serviceController) {
+            this.persistenceUnitService = persistenceUnitService;
+            this.serviceController = serviceController;
+        }
+
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        @Override
+        public ResourceReference<EntityManagerFactory> createResource() {
+            serviceController.addListener(
+                    new LifecycleListener() {
+
+                        @Override
+                        public void handleEvent(final ServiceController<?> controller, final LifecycleEvent event) {
+                            if (event == LifecycleEvent.UP) {
+                                latch.countDown();
+                                controller.removeListener(this);
+                            }
+                        }
+                    }
+            );
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                // Thread was interrupted, which we will preserve in case a higher level operation needs to see it.
+                Thread.currentThread().interrupt();
+                // rather than just returning the current EntityManagerFactory (might be null or not null),
+                // fail with a runtime exception.
+                throw new RuntimeException(e);
+            }
+            return new ResourceReference<EntityManagerFactory>() {
+                final AccessControlContext accessControlContext =
+                        AccessController.doPrivileged(GetAccessControlContextAction.getInstance());
+                EntityManagerFactory entityManagerFactory;
+
+                @Override
+                public EntityManagerFactory getInstance() {
+                    PrivilegedAction<Void> privilegedAction =
+                            new PrivilegedAction<Void>() {
+                                // run as security privileged action
+                                @Override
+                                public Void run() {
+                                    entityManagerFactory = persistenceUnitService.getEntityManagerFactory();
```

Comment:
Perhaps use a single class LazyFactory<T> and pass a Callable<T> to its constructor. One Callable impl does ^^^, the other does the object instantiation at L266. My impression is these 2 classes are the same except for the way the ultimate object is obtained.

Reply:
Switched to use a single class lazy factory https://github.com/wildfly/wildfly/pull/14539/commits/823f482fdeef5df5a67b782a8ac28f84de99c8dd (can be squashed into other change after review).